doc/source/serve/deployment-graph.md

```diff
@@ -0,0 +1,545 @@
+
+Let's put the concepts together and incrementally build a runnable DAG example including the following attributes:
+
+```{tip}
+At the end of this document we have the full and end to end executable implementation.
+```
+
+- All nodes in the deployment graph naturally forms a DAG structure.
+- A node could use or call into other nodes in the deployment graph.
+- Deployment graph has mix of class and function as nodes.
+- Same input or output can be used in multiple nodes in the DAG.
+- A node might access partial user input.
+- Control flow is used where dynamic dispatch happens with respect to input value.
+- Same class can be constructed, or function bound with different args that generates multiple distinct nodes in DAG.
+- A node can be called either sync or async.
+
++++
+
+![deployment graph](https://github.com/ray-project/images/blob/master/docs/serve/deployment_graph.png?raw=true)
+
++++
+
+### Step 1: User InputNode and preprocessor 
+
+Let's start with the first layer of DAG: Building user input to two preprocessor functions, where each function receives parts of the input data. For simplicity, we use the same existing `@serve.deployment` decorator on an async function body.
+
++++
+
+```python
+from ray import serve
+from ray.experimental.dag.input_node import InputNode
+
+@serve.deployment
+async def preprocessor(input_data: str):
+    """Simple feature processing that converts str to int"""
+    time.sleep(0.1) # Manual delay for blocking computation
+    return int(input_data)
+
+@serve.deployment
+async def avg_preprocessor(input_data):
+    """Simple feature processing that returns average of input list as float."""
+    time.sleep(0.15) # Manual delay for blocking computation
+    return sum(input_data) / len(input_data)
+
+# DAG building
+with InputNode() as dag_input:
+    preprocessed_1 = preprocessor.bind(dag_input[0])  # Partial access of user input by index
+    preprocessed_2 = avg_preprocessor.bind(dag_input[1]) # Partial access of user input by index
+```
+
++++
+
+There are two new APIs used in the DAG building stage: `InputNode()` and `bind()`.
+
+### **```InputNode()```** : User input of the graph
+
+```InputNode``` is a special singleton in DAG building that's only relevant to it's runtime call behavior. Even though all decorated classes or functions can be reused in arbitrary way to facilitate DAG building where the root DAGNode forms the graph with its children, in each deployment graph there should be one and only one InputNode used.
+
+```InputNode``` value is fulfilled and replaced by user input at runtime, therefore it takes no argument when being constructed.
+
+```InputNode``` should only be created within a context manager as syntax sugar to indicate there's one and only one `InputNode` for each graph, and will throw exception if directly created, or multiple is found.
+
+It's possible to access partial user input by index or key, if some DAGNode in the graph doesn't need the complete user input to run. 
```

Comment:
i've updated it to 

```
It's possible to access partial user input by index or key, if a node in the graph doesn't need the complete user input to run. Example: User submits `[Tensor_0, Tensor_1]` via http but a model node might only need to receive `Tensor_1` at index 1 rather than complete list to reduce intermediate data transfer.
```